### ISSUE TITLE
How the backward overflow test works

### ISSUE THREAD

**Issue by howiejayz** [Dao-AILab/flash-attention#650]
Dear Team,

I am currently developing FA2 for AMD platforms and when I investigate on the unit tests, I saw there is a new tests called backward overflow. 

After I trace through the test script, I saw the values of q_pt.grad and k_pt.grad are both all zeros after out.backward(). Is that supposed to be so? What is the main purpose for this test?

**Comment by howiejayz** [comment]
I also wonder if you are using RTN or RTZ for the unit tests to all pass?

**Comment by tridao** [comment]
These tests were added because of some [bug](https://github.com/Dao-AILab/flash-attention/issues/443) [reports](https://github.com/Dao-AILab/flash-attention/issues/334) on gradients being NaN. Multiply q and kv by some amount (3 or 5) is just to potentially trigger the logits being larger than fp16 limit that could lead to NaN without proper masking.

I'm not sure if we're using RTN or RTZ, it shouldn't matter too much. If the results are close you can increase the tolerance a bit to make it pass.

**Comment by howiejayz** [comment]
Hi dao. Thanks for clarifying! We are doing [upstream integration](https://github.com/ROCmSoftwarePlatform/flash-attention) from your version v2.0.4. In our implementation, we need to build with RTN for running UT for higher precision, and build with RTZ for benchmarks because of better performance. I am just curious how you can reach the best performance and pass all unit tests at the same time.

**Comment by tridao** [comment]
Are you referring to the fp32 -> fp16/bf16 conversion of the forward pass output? Or fp32 -> fp16/bf16 conversion of dQ, dK, dV? Or is there another conversion?
Tbh I'm not sure if I'm using RTN or RTZ, whatever is the default in Cutlass when doing fp32 -> fp16/bf16 conversion.

**Comment by tridao** [comment]
I think we're using RTN since that's what cutlass conversion function [calls](https://github.com/NVIDIA/cutlass/blob/39c6a83f231d6db2bc6b9c251e7add77d68cbfb4/include/cutlass/numeric_conversion.h#L845C45-L845C62).
